### PR TITLE
stop orders

### PR DESCRIPTION
**Pull request by saml33**
I've added the whitelist to this so we can continue testing in prod. There are a few known issues:

The table price data isn't coming through correctly. We need the next program upgrade for this to be finalized.
There's a flicker on the chart if you flip the price (save this preference to local storage) and then flip the input/output tokens
The estimated slippage values don't look right (e.g. selling MNGO has 12.8% slippage no matter the size)
We need more validation for orders that exceed the maximums for the platform

### CODE REVIEW

File: components/swap/SwapTokenChart.tsx [hunk 1]
```diff
@@ -402,6 +454,22 @@ const SwapTokenChart = () => {
...
+  //     {
+  //       price,
+  //       time: Date.now(),
+  //       inputTokenPrice: inputBank.uiPrice,
+  //       outputTokenPrice: outputBank.uiPrice,
+  //     },
+  //   ]
+  //   return item
+  // }, [flipPrices, inputBank, outputBank])
+
```
Comment (tlrsssss) [inline components/swap/SwapTokenChart.tsx:472]: Commented code. Why not show the latest price in the chart?

File: components/account/AccountTabs.tsx [hunk 2]
```diff
@@ -12,29 +12,61 @@ import useOpenPerpPositions from 'hooks/useOpenPerpPositions'
...
+  //     mangoAccount?.tokenConditionalSwaps.filter((tcs) => tcs.hasData)
+  //       ?.length || 0
+  //   tabs = [
+  //     ['balances', 0],
+  //     ['trade:trigger-orders', stopOrdersCount],
+  //     ['swap:swap-history', 0],
+  //   ]
+  // }
+  //   return tabs
+  // }, [isWhiteListed, mangoAccount])
```
Comment (tlrsssss) [inline components/account/AccountTabs.tsx:46]: can this be removed?

File: components/swap/LimitSwapForm.tsx [hunk 3]
```diff
@@ -0,0 +1,955 @@
...
+    [flipPrices, quotePrice, setFormErrors, setOrderTypeMultiplier],
+  )
+
+  // const disablePlaceOrder =
+  //   (orderType === OrderTypes.REPAY_BORROW && !hasBorrowToRepay) ||
+  //   (orderType === OrderTypes.STOP_LOSS &&
+  //     parseFloat(triggerPrice) > quotePrice) ||
+  //   (orderType === OrderTypes.TAKE_PROFIT &&
+  //     parseFloat(triggerPrice) < quotePrice) ||
+  //   amountInAsDecimal.gt(sellTokenBalance)
```
Comment (tlrsssss) [inline components/swap/LimitSwapForm.tsx:755]: Can this be removed?

File: components/swap/SwapForm.tsx [hunk 4]
```diff
@@ -364,175 +158,36 @@ const SwapForm = () => {
...
+            <>
+              {/* <div className="absolute right-4 top-4">
+                <IconButton
+                  className="text-th-fgd-3"
+                  hideBg
+                  onClick={() => setShowSettings(true)}
+                >
+                  <Cog8ToothIcon className="h-5 w-5" />
+                </IconButton>
+              </div> */}
```
Comment (tlrsssss) [inline components/swap/SwapForm.tsx:181]: Can this be removed?

File: components/swap/SwapTokenChart.tsx [hunk 5]
```diff
@@ -412,7 +480,9 @@ const SwapTokenChart = () => {
         (point) => point.time >= minTime && point.time <= maxTime,
       )
       return coingeckoData.concat(swapPoints).sort((a, b) => a.time - b.time)
+      // .concat(latestChartDataItem)
     } else return coingeckoData
+    // .concat(latestChartDataItem)
```
Comment (tlrsssss) [inline components/swap/SwapTokenChart.tsx:485]: can these be removed?

File: public/locales/ru/trade.json [hunk 6]
```diff
@@ -83,6 +89,10 @@
     "size": "Size",
     "spread": "Spread",
     "stable-price": "Stable Price",
+    "stop-limit": "Stop Limit",
+    "stop-loss": "Stop-Loss",
+    "stop-market": "Stop Market",
+    "take-profit": "Take-Profit",
```
Comment (tlrsssss) [inline public/locales/ru/trade.json:95]: Not sure a dash is needed for these

**Review by tlrsssss — CHANGES_REQUESTED** [review]
Mostly looks good just some cleanup then can merge

**Comment by tlrsssss** [comment]
small nit, noticed chart flashes when swapping tokens
https://github.com/blockworks-foundation/mango-v4-ui/assets/1254942/d76a07ac-2d8b-4abb-895f-f7ae14dbf343

**Comment by tlrsssss** [comment]
https://github.com/blockworks-foundation/mango-v4-ui/assets/1254942/b8f11202-96f9-4aa0-93f2-edd77217c39d

**Comment by tlrsssss** [comment]
https://github.com/blockworks-foundation/mango-v4-ui/assets/1254942/0f044504-7f35-4e17-a839-261bf6c7b865

small nit, noticed chart flashes when swapping tokens

**Comment by saml33** [comment]
> chartbug.mov 
> small nit, noticed chart flashes when swapping tokens
> 
> if you're flipping tokens should the chart flip as well? i think this is what happens on main

Yeah, so it saves the direction if you flip the prices on the trigger order form. It will stay in that direction even when you flip the tokens. Not sure it's necessary and I'm not sure how to get rid of this flicker

**Comment by tlrsssss** [comment]
> > chartbug.mov
> > small nit, noticed chart flashes when swapping tokens
> > if you're flipping tokens should the chart flip as well? i think this is what happens on main
> 
> Yeah, so it saves the direction if you flip the prices on the trigger order form. It will stay in that direction even when you flip the tokens. Not sure it's necessary and I'm not sure how to get rid of this flicker

I suspect that keeping it in the direction once flipping it, is related to that flicker. If it's not necessary, can try getting rid of it to see if that solves the flicker.

**Comment by saml33** [comment]
> > > chartbug.mov
> > > small nit, noticed chart flashes when swapping tokens
> > > if you're flipping tokens should the chart flip as well? i think this is what happens on main
> > 
> > 
> > Yeah, so it saves the direction if you flip the prices on the trigger order form. It will stay in that direction even when you flip the tokens. Not sure it's necessary and I'm not sure how to get rid of this flicker
> 
> I suspect that keeping it in the direction once flipping it, is related to that flicker. If it's not necessary, can try getting rid of it to see if that solves the flicker.

Yeah, it's because of the flipPrices boolean. I think I'll get rid of it. Adding the latestPrice (added back now) also causes a flicker as it needs flipPrices. Maybe it will work if I put it here: fetchChartData

**Review by tlrsssss — APPROVED** [review]
looks good, just need to fix the yarn dedupe

**Pull request merged**